src/main/java/hudson/plugins/ec2/SlaveTemplate.java

```diff
@@ -1210,20 +1210,31 @@ List<InstanceInfo> createNewInstances(AmazonEC2 ec2, RunInstancesRequest riReque
         if (isUsingDynamicSshKeys()) {
             // no static ssh key defined, so submit riRequest.maxCount requests
             // each with a unique keypair (instead of a single request for multiple identical instances)
+            LOGGER.fine(() -> "no static ssh credential configured, will use dynamic ssh keys instead");
             List<InstanceInfo> instances = new ArrayList<>();
             int maxRequested = riRequest.getMaxCount();
             riRequest.setMaxCount(1);
             for (int i=0;i < maxRequested; i++) {
                 String keyName = "cloudbees-" + UUID.randomUUID();
                 KeyPair keyPair = ec2.createKeyPair(new CreateKeyPairRequest(keyName)).getKeyPair();
+                LOGGER.fine("created new dynamic keypair " + keyPair.getKeyName());
                 riRequest.setKeyName(keyPair.getKeyName());
                 Instance instance = ec2.runInstances(riRequest).getReservation().getInstances().get(0);
                 instances.add(new InstanceInfo(instance, keyPair));
             }
             return instances;
         } else {
             // using a static ssh key
-            return InstanceInfo.fromInstances(ec2.runInstances(riRequest).getReservation().getInstances());
+            List<InstanceInfo> instances = InstanceInfo.fromInstances(ec2.runInstances(riRequest).getReservation().getInstances());
+            instances.stream().forEach(instance -> {
+                try {
+                    LOGGER.fine(() -> "static ssh credential configured, retrieving keypair and setting it in instance");
+                    instance.setKeypair(getParent().resolveKeyPair());
```

Comment:
Seems off. The `InstanceInfo.keypair` should have been populated in its constructor by `fromInstances`.

Reply:
Apparently i missed removing that after implementing the requested change :facepalm: